cpp/include/cudf/detail/gather.cuh

```diff
@@ -678,9 +679,24 @@ std::unique_ptr<table> gather(table_view const& source_table,
                           return col.nullable();
                         });
   if (nullable) {
-    auto const op = bounds_policy == out_of_bounds_policy::NULLIFY ? gather_bitmask_op::NULLIFY
-                                                                   : gather_bitmask_op::DONT_CHECK;
-    gather_bitmask(source_table, gather_map_begin, destination_columns, op, stream, mr);
+    auto const has_nulls =
+      bounds_policy == out_of_bounds_policy::NULLIFY ||
+      std::any_of(
+        source_table.begin(), source_table.end(), [](auto const& col) { return col.has_nulls(); });
```

Comment:
This actually already exists as [`cudf::has_nulls`](https://github.com/rapidsai/cudf/blob/branch-23.12/cpp/include/cudf/table/table_view.hpp#L318). However, I suspect that the reason that tests are failing is that what you really mean to check is whether there are nested nulls, which also exists as [`cudf::has_nested_nulls](https://github.com/rapidsai/cudf/blob/branch-23.12/cpp/include/cudf/table/table_view.hpp#L329)? That may at least account for the struct cases.

Reply:
That did solve `STRUCT_TEST` failures but these tests still fail for me locally:
```[  FAILED  ] JoinTest.Repro_StructsWithoutNullsPushedDown```

```[  FAILED  ] 24 tests, listed below:
[  FAILED  ] groupby_median_test/0.zero_valid_values, where TypeParam = signed char
[  FAILED  ] groupby_median_test/0.null_keys_and_values, where TypeParam = signed char
[  FAILED  ] groupby_median_test/1.zero_valid_values, where TypeParam = short
[  FAILED  ] groupby_median_test/1.null_keys_and_values, where TypeParam = short
[  FAILED  ] groupby_median_test/2.zero_valid_values, where TypeParam = int
[  FAILED  ] groupby_median_test/2.null_keys_and_values, where TypeParam = int
[  FAILED  ] groupby_median_test/3.zero_valid_values, where TypeParam = long
[  FAILED  ] groupby_median_test/3.null_keys_and_values, where TypeParam = long
[  FAILED  ] groupby_median_test/4.zero_valid_values, where TypeParam = float
[  FAILED  ] groupby_median_test/4.null_keys_and_values, where TypeParam = float
[  FAILED  ] groupby_median_test/5.zero_valid_values, where TypeParam = double
[  FAILED  ] groupby_median_test/5.null_keys_and_values, where TypeParam = double
[  FAILED  ] groupby_quantile_test/0.zero_valid_values, where TypeParam = signed char
[  FAILED  ] groupby_quantile_test/0.null_keys_and_values, where TypeParam = signed char
[  FAILED  ] groupby_quantile_test/1.zero_valid_values, where TypeParam = short
[  FAILED  ] groupby_quantile_test/1.null_keys_and_values, where TypeParam = short
[  FAILED  ] groupby_quantile_test/2.zero_valid_values, where TypeParam = int
[  FAILED  ] groupby_quantile_test/2.null_keys_and_values, where TypeParam = int
[  FAILED  ] groupby_quantile_test/3.zero_valid_values, where TypeParam = long
[  FAILED  ] groupby_quantile_test/3.null_keys_and_values, where TypeParam = long
[  FAILED  ] groupby_quantile_test/4.zero_valid_values, where TypeParam = float
[  FAILED  ] groupby_quantile_test/4.null_keys_and_values, where TypeParam = float
[  FAILED  ] groupby_quantile_test/5.zero_valid_values, where TypeParam = double
[  FAILED  ] groupby_quantile_test/5.null_keys_and_values, where TypeParam = double```